### PR TITLE
Implement a fast (lcg) random generator for Python SyntheticSource

### LINKED ISSUE #25944: [Task]: Improve the performance of Python Synthetic Source
### What needs to happen?

It is found that the cost of generating synthetic source is as expensive as write to sink in a Python IO performance test https://github.com/apache/beam/issues/19084#issuecomment-1343373709 . This prevents the benchmark from reporting accurate performance data.

Ran a pipeline with synthetic source only, cloud profile shows

- assign random seed alone costs 30% of total cpu time
- generate bytes costs 20% of total cpu time

image
 
This is because Python built in random generator uses a Mersenne Twister with fairly large state ((doc)[https://docs.python.org/3/library/random.html]), thus assigning seed is slow. Generating bytes is also slow as it involves many memory allocations. In contrast, Java built in random generator (used by Java SDK's synthetic source) uses a linear congruential generator (LCG) by Donald Knuth ((doc)[https://docs.oracle.com/javase/8/docs/api/java/util/Random.html]) which is way faster.

I compared the performance between builtin generating random bytes and cythonized LCG implemenration, generating 1M random bytes of 1024 bytes. The latter shows more than 10 x performance gain (run time 10 s / < 1 s). This doubles the performance of synthetic pipeline. We should be able to switch to the LCG

Once this is done Python synthetic pipeline has minimum cost of generating bytes themselves and can then be used to benchmarking the peformance of SDF.

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by Abacn**
Fixes #25944

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
run seed job

**Comment by Abacn** [comment]
Run Python TextIO Performance Test

**Comment by Abacn** [comment]
Python xlang KafkaIO Performance Test

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Run Python xlang KafkaIO Performance Test

**Comment by Abacn** [comment]
Run Python TextIO Performance Test

**Comment by Abacn** [comment]
Python xlang KafkaIO Performance Test

**Comment by Abacn** [comment]
Run Python xlang KafkaIO Performance Test

**Comment by Abacn** [comment]
The change is effective in performance tests:

Previous:
image


Now:
image

**Comment by Abacn** [comment]
run seed job

**Comment by Abacn** [comment]
Run Load Tests Python Combine Dataflow Batch

**Comment by Abacn** [comment]
Run Load Tests Python GBK Dataflow Streaming

**Comment by Abacn** [comment]
run seed job

**Comment by Abacn** [comment]
Run Load Tests Python Combine Dataflow Batch

**Comment by Abacn** [comment]
Run Load Tests Python GBK Dataflow Batch

**Comment by Abacn** [comment]
Change is effective to load test:

On master (https://ci-beam.apache.org/job/beam_LoadTests_Python_Combine_Dataflow_Batch/1405/console)

INFO:apache_beam.testing.load_tests.load_test_metrics_utils:Metric: python_dataflow_batch_combine_1_runtime Value: 621

Now (https://ci-beam.apache.org/job/beam_LoadTests_Python_Combine_Dataflow_Batch_PR/20/console)

INFO:apache_beam.testing.load_tests.load_test_metrics_utils:Metric: python_dataflow_batch_combine_1_runtime Value: 352

This is a 43% improvement. Originally _all_ Python load tests actually are testing the performance of SyntheticSource as it costs mosts of the run time.

**Comment by Abacn** [comment]
Haven't change Flink load test because they do not use sdk_location but run on the latest sdk container snapshot. Only when this change gets merged into master can see difference on flink load test

**Comment by Abacn** [comment]
R: @tvalentyn @AnandInguva

**Comment by tvalentyn** [comment]
Nice change and blogpost! how does this implementation compare with numpy's generators in terms of performance?

**Comment by Abacn** [comment]
> Nice change and blogpost! how does this implementation compare with numpy's generators in terms of performance?

Thanks for taking a look! I also tested numpy's default generator (PCG64) and it seems much slower than builtin

Code:

```python
num_loop = 1_000_000
length = 1024

startt = time.time()
    for idx in range(num_loop):
        gen = np.random.default_rng(idx)
        c = gen.bytes(length)
    print("Use numpy: ", time.time() - startt)
```

Example run: Use numpy:  34.7236704826355

For comparison, the cython implementation takes 1 s; builtin takes 10 s.

If I do not set seed every time, the run time is sth like "Use numpy:  10.303484201431274" similar to builtin, but for numpy it seems not a seed method setting seed of an existing generator available

File: sdks/python/apache_beam/testing/synthetic_pipeline.py [hunk 1]
```diff
@@ -83,7 +91,14 @@ def bytes(self, length):
     return self.getrandbits(length * 8).to_bytes(length, sys.byteorder)
 
 
-Generator = _Random
+def get_generator(seed: Optional[int] = None, algorithm: Optional[str] = None):
+  if algorithm is None or algorithm == 'builtin':
+    return _Random(seed)
+  else:
```
Comment (tvalentyn) [inline sdks/python/apache_beam/testing/synthetic_pipeline.py:97]: Let's check or assert that algorithm=='lcg'?

File: sdks/python/apache_beam/testing/test_utils_test.py [hunk 2]
```diff
@@ -95,6 +95,43 @@ def test_cleanup_topics(self):
...
+    self.assertEqual(generator.rand_bytes(10), b'\xa6\x8fW\xcb\xb1\xa88]dP')
+
+  def test_generator_seed_jdk_results(self):
+    generator = self.Generator()
+    generator.seed_jdk(0)
+    self.assertEqual(generator.next_int(), -1155484576)
+
+    generator.seed_jdk(1)
+    # the first next_int after seed_jdk(1) is close to seed_jdk(0)
+    self.assertEqual(generator.next_int(), -1155869325)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/testing/test_utils_test.py:119]: just curious (you probably verified this) - does the value still match the JDK value after some large number of iteration? no need to make a test for this.

**Comment by tvalentyn** [comment]
> 
This is a 43% improvement. Originally all Python load tests actually are testing the performance of SyntheticSource as it costs mosts of the run time.

sounds like we should consider using pregenerated datasets...

**Review by tvalentyn — APPROVED** [review]

File: sdks/python/apache_beam/testing/test_utils_test.py (hunk 2 above)
Comment (Abacn) [reply sdks/python/apache_beam/testing/test_utils_test.py:119]: Just did some more test:

Python code:

```python

from apache_beam.testing.synthetic_pipeline import get_generator

if __name__ == '__main__':
  gen = get_generator(algorithm='lcg')
  gen.seed_jdk(1)
  for i in range(10):
    print(gen.next_int())
```

Java code:
```java

public static void main(String[] argv) {
  Random rand = new Random(1);
  for (int i = 0; i < 10; ++i) {
    System.out.println(rand.nextInt());
  }
}
```

Both has result
```
-1155869325
431529176
1761283695
1749940626
892128508
155629808
1429008869
-1465154083
-138487339
-1242363800
```

File: sdks/python/apache_beam/testing/synthetic_pipeline.py (hunk 1 above)
Comment (Abacn) [reply sdks/python/apache_beam/testing/synthetic_pipeline.py]: added check

**Pull request merged**